src/fsharp/TastPickle.fs

```diff
@@ -1539,6 +1555,7 @@ let u_tyar_specs = (u_list u_tyar_spec)
 
 let _ = fill_p_ty (fun ty st ->
     let ty = stripTyparEqns ty
+    let ty = if isInByrefTy st.oglobals ty && st.isStructThisArgPos then destByrefTy st.oglobals ty else ty
```

Comment:
Why are we getting the destination of the `inref` here?

Reply:
Oh crap, yes, this is wrong, It should be mkByrefTy (destByrefTy st.oglobals ty), i.e. convert the `inref` to a `byref`.  

In a way this shows how much the stored `this` type doesn't actually affect things in the compiler, but I'll definitely adjust to get it right.